src/java/org/apache/cassandra/metrics/ClientMetrics.java

```diff
@@ -118,6 +193,31 @@ public long getCount()
         authSuccess = registerMeter("AuthSuccess");
         authFailure = registerMeter("AuthFailure");
 
+        // For each of SSL, non-SSL register a gauge:
+        encryptedConnectedNativeClients = registerGauge(new DefaultNameFactory("Client", "Encrypted"), CONNECTED_NATIVE_CLIENTS, () -> countConnectedClients((ServerConnection::isSSL)));
+        unencryptedConnectedNativeClients = registerGauge(new DefaultNameFactory("Client", "Unencrypted"), CONNECTED_NATIVE_CLIENTS, () -> countConnectedClients(((ServerConnection connection) -> !connection.isSSL())));
+
+        // for each supported authentication mode, register a meter for success and failures.
+        final IAuthenticator authenticator = DatabaseDescriptor.getAuthenticator();
+        for (String mode : authenticator.getSupportedAuthenticationModes())
+        {
+            String capitalizedMode = StringUtils.capitalize(mode);
+            MetricNameFactory factory = new DefaultNameFactory("Client", capitalizedMode);
+            authSuccessByMode.put(mode, registerMeter(factory, AUTH_SUCCESS));
+            authFailureByMode.put(mode, registerMeter(factory, AUTH_FAILURE));
+
+            Gauge<Integer> clients = registerGauge(factory, CONNECTED_NATIVE_CLIENTS, () -> countConnectedClients((ServerConnection connection) -> {
+                final ClientState state = connection.getClientState();
```

Comment:
Why is `state` final and `user` is not? Why doe it have to be final?

Reply:
this might be all simplified to

                AuthenticatedUser user = connection.getClientState().getUser();
                return Optional.ofNullable(user).map(u -> mode.equals(u.getMode())).orElse(false);